### PR TITLE
OOI inputs

### PR DESCRIPTION
**Pull request by Molkree**
This adds input parameters to the OOI workflow/script. Most important are node, time start/end, segment length, but also output directory and NFFT (like in Orcasound).

Want to add some tests first (mainly for parsing inputs).

Part of #8 #9 #35 for the OOI side.

### CODE REVIEW

**Comment by Molkree** [comment]
I'm not sure tests run so long, 5min43sec for 1min+5min+5min of data 🤔
But it works so I guess it's fine

File: .github/workflows/ooi_processing.yml [hunk 1]
```diff
@@ -3,10 +3,30 @@ name: OOI processing
...
+      start_time:
+        description: 'start_time formatted as Y-m-dTH-M-S'
+        required: false
+      end_time:
+        description: 'end_time formatted as Y-m-dTH-M-S'
+        required: false
+      segment_length:
+        description: 'Segment length in minutes'
+        required: false
+        default: 5
```
Comment (valentina-s) [inline .github/workflows/ooi_processing.yml:20]: Can time be smaller than a minute? Using fractions?  If we want spectrograms on a whale call length, it is usually only a few seconds?

File: tests/test_ooi.py [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+from ooi_processing import save_ooi_spectrograms
+
+
+@pytest.mark.parametrize(
+    "start_time_str, end_time_str, segment_length_float, node, output_dir, expected_files_count",
+    [
+        ("2017-03-10T00-00-00", "2017-03-10T00-05-00", None, None, None, 1),
+        ("2017-03-10T00-00-00", "2017-03-10T00-05-00", 1, None, None, 5),
+        ("2017-03-10T00-00-00", "2017-03-10T00-05-00", 1, None, "long/path/", 5),
```
Comment (valentina-s) [inline tests/test_ooi.py:16]: Test end_time before start_time.

File: tests/test_spectrograms.py [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+
+import pytest
+
+from create_spectrogram import create_spec_name, save_spectrogram
+
+
+@pytest.mark.parametrize(
+    "wav_name, output_dir, expected",
+    [
+        ("2021-01-01T00-00-00-000.wav", None, "2021-01-01T00-00-00-000.png"),
```
Comment (valentina-s) [inline tests/test_spectrograms.py:12]: test when file has no .wav extension.

File: .github/workflows/ooi_processing.yml (hunk 1 above)
Comment (Molkree) [reply .github/workflows/ooi_processing.yml:20]: Yes, it is actually a float, so can be really small, I should probably indicate that in the description!

File: tests/test_spectrograms.py [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+    "wav_name, plot_path",
+    [
+        ("tests/ooi.wav", None),
+        ("tests/ooi.wav", "tests/path/to/output/ooi.png"),
+        ("tests/orcasound.wav", "tests/path/to/output/orcasound.png"),
+    ],
+)
+def test_save_spectrogram(wav_name, plot_path):
+    spec_path = save_spectrogram(wav_name, plot_path)
+    assert os.path.isfile(spec_path)
```
Comment (valentina-s) [inline tests/test_spectrograms.py:44]: check something about the content of the file: so the plot is not empty. Compare with the one you have (is the plot deterministic?)

File: create_spectrogram.py [hunk 5]
```diff
@@ -28,12 +45,12 @@ def save_spectrogram(input_wav, plot_path=None, NFFT=256):
         `plot_path`: Path to the output spectrogram file. Default is `input_wav` with .png extension.
         `NFFT`: The number of data points used in each block for the FFT. A power 2 is most efficient.
     Returns:
-        None
+        Path to the spectrogram.
     """
     samplerate, data = wavfile.read(input_wav)
```
Comment (valentina-s) [inline create_spectrogram.py:50]: with try-except you can catch io error.

File: create_spectrogram.py [hunk 6]
```diff
@@ -28,12 +45,12 @@ def save_spectrogram(input_wav, plot_path=None, NFFT=256):
...
     noverlap = NFFT // 2 if NFFT <= 128 else 128
 
-    title = input_wav.removesuffix(".wav")
+    title = input_wav.replace(".wav", "")
```
Comment (valentina-s) [inline create_spectrogram.py:53]: with try-except you can catch io error.

File: ooi_processing.py [hunk 7]
```diff
@@ -1,35 +1,124 @@
...
+        `start_time`: `datetime.datetime`
+        `end_time`: `datetime.datetime`
+        `segment_length`: `datetime.timedelta`
+        `node`: One of the preset OOI nodes.
+        `output_dir`: Path to the output directory.
+        `NFFT`: The number of data points used in each block for the FFT. A power 2 is most efficient.
+    Returns:
+        None
+    """
+    while start_time < end_time:
```
Comment (valentina-s) [inline ooi_processing.py:33]: For now this approach is ok, in the future possibly a for loop will help with parallelization (maybe if one gets a runner with more cores)

**Comment by valentina-s** [comment]
@Molkree PR looks good! I added a few comments inline.

File: tests/test_spectrograms.py (hunk 4 above)
Comment (Molkree) [reply tests/test_spectrograms.py:46]: @valentina-s, how would I compare them? The plot is deterministic I think, yes. I thought about using [`filecmp.cmp`](https://docs.python.org/3/library/filecmp.html#:~:text=the%20following%20functions%3A-,filecmp.cmp,-(f1%2C%20f2%2C%20shallow)).

Comment (Molkree) [reply tests/test_spectrograms.py]: For now I've added `filecmp.cmp`, let's see how it works

File: create_spectrogram.py (hunk 6 above)
Comment (Molkree) [reply create_spectrogram.py]: I've changed this part a bit so it now looks like this `title = path.splitext(path.basename(input_wav))[0]` but what io error can happen here?

File: ooi_processing.py (hunk 7 above)
Comment (Molkree) [reply ooi_processing.py:33]: Good idea! I've opened an issue to track it #46

File: create_spectrogram.py (hunk 6 above)
Comment (valentina-s) [reply create_spectrogram.py]: I think I referred to the line above which is reading the file: `wavfile.read`

Comment (Molkree) [reply create_spectrogram.py]: Okay, I see but why would we want to catch here? If we can't read the file I don't see what else we could do.

Comment (valentina-s) [reply create_spectrogram.py]: I see that the error gets logged anyway.

**Pull request merged**